controller/node_controller.go

```diff
@@ -1441,3 +1446,107 @@ func (nc *NodeController) createSnapshotMonitor() (mon monitor.Monitor, err erro
 
 	return mon, nil
 }
+
+func (nc *NodeController) syncReplicaEvictionRequested(node *longhorn.Node, kubeNode *corev1.Node) error {
+	log := getLoggerForNode(nc.logger, node)
+	node.Status.AutoEvicting = false
+	nodeDrainPolicy, err := nc.ds.GetSettingValueExisted(types.SettingNameNodeDrainPolicy)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %v setting", types.SettingNameNodeDrainPolicy)
+	}
+
+	type replicaToSync struct {
+		*longhorn.Replica
+		syncReason string
+	}
+	replicasToSync := []replicaToSync{}
+
+	for diskName, diskSpec := range node.Spec.Disks {
+		diskStatus := node.Status.DiskStatus[diskName]
+		for replicaName := range diskStatus.ScheduledReplica {
+			replica, err := nc.ds.GetReplica(replicaName)
+			if err != nil {
+				return err
+			}
+			shouldEvictReplica, reason, err := nc.shouldEvictReplica(node, kubeNode, &diskSpec, replica,
+				nodeDrainPolicy)
+			if err != nil {
+				return err
+			}
+			if replica.Spec.EvictionRequested != shouldEvictReplica {
+				replica.Spec.EvictionRequested = shouldEvictReplica
+				replicasToSync = append(replicasToSync, replicaToSync{replica, reason})
+			}
+
+			if replica.Spec.EvictionRequested && !node.Spec.EvictionRequested && !diskSpec.EvictionRequested {
+				// We don't consider the node to be auto evicting if eviction was manually requested.
+				node.Status.AutoEvicting = true
+			}
+		}
+	}
+
+	for _, replicaToSync := range replicasToSync {
+		replicaLog := log.WithField("replica", replicaToSync.Name).WithField("disk", replicaToSync.Spec.DiskID)
+		if replicaToSync.Spec.EvictionRequested {
+			replicaLog.Infof("Requesting replica eviction")
+			if _, err := nc.ds.UpdateReplica(replicaToSync.Replica); err != nil {
+				replicaLog.Warn("Failed to request replica eviction, will enqueue then resync node")
+				nc.enqueueNode(node)
```

Comment:
NIT: should we enqueue with a rate-limited speed? I am worried that we might hit an unresolvable error and keep enquiring under a tight loop which might affect performance.

Reply:
Added rate limiting, but ran out of time to test. I'll do this first thing after the break and we can merge then.